pkg/logs/decoder/gdecoder.go

```diff
@@ -0,0 +1,104 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package decoder
+
+import "time"
+
+// GDecoder should replace current Decoder object.
+// Decoder encapsulates the logic for processing logs. The communication is done by
+// InputChan and OutputChan channels. Logs are received by byte arrays as contents of
+// Inputs and sent with extra formatting and resizing as contents of Outputs.
+//
+// The main components of decoder logic are: LineGenerator, LineHandlerScheduler,
+// LineHandler, LineTruncator.
+//
+// The data flows like this:
+// Input -> LineGenerator(cache and convert) -> RichLine -> LineHandlerScheduler ->
+// LineHandler -> RichLine(s) -> LineTruncator -> Output
+//
+// The main job of LineGenerator is to read Inputs byte by byte forms Line with extra
+// information and send to downstream.
+// LineHandlerScheduler keeps a timer to make sure LineHandlers clean up their caches
+// before terminating.
+// LineHandler handles the Line according to specifications.
+// LineTruncator finally apply the truncation rules that are accumulated by the upstream.
+type GDecoder struct {
+	InputChan     chan *Input
+	OutputChan    chan *Output
+	lineGenerator LineGenerator
+}
+
+// Start prepares routine for consuming inputs.
+func (d *GDecoder) Start() {
+	d.lineGenerator.Start()
+}
+
+// Stop closes the input channel in order to terminate the decoding job.
+func (d *GDecoder) Stop() {
+	close(d.InputChan)
+}
+
+const defaultMaxDecodeLength = 1024 * 1024 // 1MB
+const defaultMaxSendLength = 256 * 1024    // 256KB
+
+// LineHandlerScheduler contains the logic to process / reads input channel with
+// a timer. It's to make sure, when timer times out, to clean up the cache of
+// line handler, if exists, before termination.
+type LineHandlerScheduler struct {
```

Comment:
This Handler does not schedule anything

Reply:
????

Reply:
Yes, this one is a nit pick, but we can discuss this in the final PR, I expect a Scheduler to have a schedule method.